File: app/src/main/java/me/shreyasr/chatse/event/presenter/message/MessageEvent.java
package me.shreyasr.chatse.event.presenter.message;

import android.support.annotation.NonNull;

import me.shreyasr.chatse.event.ChatEvent;

public class MessageEvent implements Comparable<MessageEvent> {

    public final String content;
    public final long timestamp;
    public final long id;
    public final long userId;
    public final String userName;
    public final long roomId;
    public final String roomName;
    public final int messageId;
    public final long parentId;
    public final int editCount;
    public final boolean onebox;
    public final String onebox_type;
    public final String onebox_content;
    MessageEvent previous;
    public MessageEvent(ChatEvent baseEvent) {
        this.content = baseEvent.contents;
        this.timestamp = baseEvent.time_stamp;
        this.id = baseEvent.id;
        this.userId = baseEvent.user_id;
        this.userName = baseEvent.user_name;
        this.roomId = baseEvent.room_id;
        this.roomName = baseEvent.room_name;
        this.messageId = baseEvent.message_id;
        this.parentId = baseEvent.parent_id;
        this.editCount = baseEvent.message_edits;
        this.onebox = baseEvent.message_onebox;
        this.onebox_type = baseEvent.onebox_type;
        this.onebox_content = baseEvent.onebox_content;
    }

    public boolean isDeleted() {
        return content == null;
    }

    public boolean isEdited() {
        return (previous != null || editCount != 0) && !isDeleted();
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof MessageEvent)) return false;
        MessageEvent event = (MessageEvent) o;
        return this.messageId == event.messageId;
    }

    @Override
    public int compareTo(@NonNull MessageEvent other) {
        if (this.equals(other)) {
            return 0;
        }
        return -Long.valueOf(timestamp).compareTo(other.timestamp);
    }
}

File: app/src/main/java/me/shreyasr/chatse/event/presenter/message/MessageEvent.kt
package me.shreyasr.chatse.event.presenter.message

import me.shreyasr.chatse.event.ChatEvent

class MessageEvent(baseEvent: ChatEvent) : Comparable<MessageEvent> {
    var content: String?
    val timestamp: Long
    val id: Long
    var userId: Long
    var userName: String
    val roomId: Long
    val roomName: String
    val messageId: Int
    val parentId: Long
    val editCount: Int
    val onebox: Boolean
    val onebox_type: String
    var message_stars: Int
    val onebox_content: String
    var onebox_extra: String
    var message_starred: Boolean
    internal var previous: MessageEvent? = null

    init {
        this.content = baseEvent.contents
        this.timestamp = baseEvent.time_stamp
        this.id = baseEvent.id.toLong()
        this.userId = baseEvent.user_id.toLong()
        this.userName = baseEvent.user_name
        this.roomId = baseEvent.room_id.toLong()
        this.roomName = baseEvent.room_name
        this.messageId = baseEvent.message_id
        this.parentId = baseEvent.parent_id.toLong()
        this.editCount = baseEvent.message_edits
        this.message_stars = baseEvent.message_stars
        this.message_starred = baseEvent.message_starred
        this.onebox = baseEvent.message_onebox
        this.onebox_type = baseEvent.onebox_type
        this.onebox_content = baseEvent.onebox_content
        this.onebox_extra = baseEvent.onebox_extra
    }

    val isDeleted: Boolean
        get() = content == null || content == ""

    val isEdited: Boolean
        get() = (previous != null || editCount != 0) && !isDeleted

    override fun equals(other: Any?): Boolean {
        if (other !is MessageEvent) return false
        val event = other as MessageEvent?
        return this.messageId == event?.messageId
    }

    override fun compareTo(other: MessageEvent): Int {
        if (this == other) {
            return 0
        }
        return other.timestamp.compareTo(this.timestamp)
    }

    override fun hashCode(): Int {
        var result = content?.hashCode() ?: 0
        result = 31 * result + timestamp.hashCode()
        result = 31 * result + id.hashCode()
        result = 31 * result + userId.hashCode()
        result = 31 * result + userName.hashCode()
        result = 31 * result + roomId.hashCode()
        result = 31 * result + roomName.hashCode()
        result = 31 * result + messageId
        result = 31 * result + parentId.hashCode()
        result = 31 * result + editCount
        result = 31 * result + onebox.hashCode()
        result = 31 * result + onebox_type.hashCode()
        result = 31 * result + onebox_content.hashCode()
        result = 31 * result + (previous?.hashCode() ?: 0)
        return result
    }
}
